File: kawazoe/q2029/tyama_codeiq2029.ex
#!/usr/bin/env elixir
defmodule StepUpSum do
	def calc(r,i,n) do
		x = n-div(i*(i-1),2)
		if x>=i do
			if rem(x,i)==0 do
				calc(r+div(x,i),i+1,n)
			else
				calc(r,i+1,n)
			end
		else
			r
		end
	end
	def main() do
		s=IO.gets('')
		if s != :eof do
			n=elem Integer.parse(s),0
			if n != 0 do
				IO.puts calc(0,2,n)
				main
			end
		end
	end
end
StepUpSum.main

File: kawazoe/q2029/tyama_codeiq2029.erl
#!/usr/bin/env escript
-module(prog).
-export([main/0]).

calc(R,I,N) ->
	X=N - (I*(I-1) div 2),
	if
		X>=I -> if
			X rem I==0 -> calc(R+(X div I),I+1,N);
			true -> calc(R,I+1,N)
		end;
		true -> R
	end.

main() ->
	case io:get_line(standard_io,"") of 
		eof -> true;
		X ->
			N=list_to_integer(string:strip(X,right,10)),
			if
				N>0 -> io:format("~w\n",[calc(0,2,N)]),main();
				true -> true
			end
	end.

main(_) -> main().